src/riak_kv_bucket.erl

```diff
@@ -339,40 +370,50 @@ validate_create_consistent_props(Invalid, New) ->
 %% validate the correctness of the type, since it assumes that was
 %% done at creation, only that it is either the same as existing or
 %% not present.
-
 -spec validate_create_dt_props(props()) -> {props(), props(), errors()}.
 validate_create_dt_props(New) ->
-    validate_create_dt_props(proplists:get_value(datatype, New), New).
-
-%% @private validate the datatype, if present
--spec validate_create_dt_props(undefined | atom(), props()) -> {props(), props(), errors()}.
-validate_create_dt_props(undefined, New) ->
-    {New, [], []};
-validate_create_dt_props(DataType, New) ->
-    Unvalidated = lists:keydelete(datatype, 1, New),
-    Mod = riak_kv_crdt:to_mod(DataType),
-    case lists:member(Mod, ?V2_TOP_LEVEL_TYPES) of
-        true ->
-            validate_create_dt_props(Unvalidated, [{datatype, DataType}], []);
-        false ->
-            Err = lists:flatten(io_lib:format("~p not supported for bucket datatype property", [DataType])),
-            validate_create_dt_props(Unvalidated, [], [{datatype, Err}])
-    end.
-
-%% @private validate the boolean property, if `datatype' was present,
-%% require `allow_mult=true' even if `datatype' was invalid, as we
-%% assume the user meant to create `datatype' bucket
--spec validate_create_dt_props(props(), props(), errors()) -> {props(), props(), errors()}.
-validate_create_dt_props(Unvalidated0, Valid, Invalid) ->
-    Unvalidated = lists:keydelete(allow_mult, 1, Unvalidated0),
-    case allow_mult(Unvalidated0) of
-        true ->
-            {Unvalidated, [{allow_mult, true} | Valid], Invalid};
+    case proplists:get_value(datatype, New) of
+        undefined -> {New, [], []};
         _ ->
-            Err = io_lib:format("Data Type buckets must be allow_mult=true", []),
-            {Unvalidated, Valid, [{allow_mult, Err} | Invalid]}
+            {Unvalidated, Valid, Errors, _} =
+                lists:foldl(fun validate_dt_props/2, {New, [], [], []},
+                            ?DT_PROPS_CHECK_CREATE),
+            {Unvalidated, Valid, Errors}
     end.
 
+%% @private generalized validation function for checking multiple datatype
+%%          properties.
+validate_dt_props(PropCheck, {Unvalidated0, Valid, Errors, Existing}) ->
+    {Prop, Fun, Err0} = PropCheck,
+    PropVal = proplists:get_value(Prop, Unvalidated0),
+    ExistingVal = proplists:get_value(Prop, Existing),
+    Unvalidated1 = lists:keydelete(Prop, 1, Unvalidated0),
+    FunVal = case Prop == datatype of
```

Comment:
If you're going to use a case rather than if here, isn't 
 
    case Prop of
        datatype -> …

cleaner?